### PR TITLE
feat(langchain): add support for streamed calls

### PR DESCRIPTION
**Pull request by sabrenner**
## What does this PR do?
Adds support for `.(a)stream(...)` calls on LangChain LCEL chains, chat models, and completion (LLM) models. It accomplishes this by:

1. starting the span when the `stream` function is called
2. as the stream is consumed, add the chunk to a list of chunks
3. when the stream is consumed, finish the span, and concatenate the results into an output string

There is one caveat to this, where it's possible the last step in a chain's `stream` call is a `JSONOutputParser`. In this case, the stream is already concatenated for us, and we use that result instead. 

A few additional notes:
- The async versions, `astream`, aren't actually async functions, they just return async generators. This is reflected in shared patching functions and test snapshots.
- This whole process is accomplished through a `shared_stream`, which returns a compatible iterable for sync and async stream managers. It utilizes `on_span_started` and `on_span_finished` functions to coordinate tags to add in the different cases, such as chain, chat, or llm.
- Since the `stream` methods do not invoke the underlying `generate` methods we trace, there's not easy path for code re-use. Thus, I tried to mimic the tags we add for the relevant `chain.invoke`, `model.generate`, and `llm.invoke` patched functions.

**Note:** The version of `vcrpy` which we pinned for reduced flakiness did not like streamed calls when using the LangChain library. As such, I introduced a new fixture that returns a stub for the HTTP transport the `OpenAI` client uses. Then, the client with that transport specified can be used on the `langchain_openai` instance. This approach uses text files with just the response data, which is why the "cassettes" added for the tests written aren't the usual `yaml` format.

## Limitations
There are a couple of limitations that need to be noted:
1. Because of [upstream issues](https://github.com/langchain-ai/langchain/issues/26719) with langchain not handling indices returned from streamed chunks from certain partner libraries (primarily OpenAI), we cannot properly aggregate streamed multiple choices by index onto tags.
2. We are not providing direct tracing support for `astream_events` in this PR. Although it uses `astream` under the hood, and it is tested that `astream_events` does not conflict with the support added in this PR, official tracing support will be added in a follow-up PR
3. Additionally, submitting streamed responses to LLM Observability will be in a direct follow-up PR as well.

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

### CODE REVIEW

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [inline ddtrace/contrib/internal/langchain/patch.py:1034]: I added all of this logic because, unlike `model.generate`, which just takes in a list of `BaseMessage` types (ie `HumanMessage`, `SystemMessage`, etc.), `stream` can take in:

1. a single string
2. a single dict
3. a list of strings
4. a list of dicts
5. a list of `BaseMessage` types
6. a `PromptValue` type, which has a `messages` property of `BaseMessage` types

Do we care about all of these different types? Or, do we just want to listify and `str` each element (this logic would also carry over to LLMObs spans, future PR)? It would make the code a lot simpler, but maybe the view of each tag not as nice.

File: ddtrace/contrib/internal/langchain/patch.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1064]:
```suggestion
                        "langchain.request.messages.%d.content" % (message_idx), integration.trunc(str(message))
```
Just to be safe, otherwise set_tag_str() will always crash

File: ddtrace/contrib/internal/langchain/patch.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1067]: I think we can try reducing the nested indents here

File: ddtrace/contrib/internal/langchain/patch.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1075]: Same here

File: ddtrace/contrib/internal/langchain/patch.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1079]: Is usage metadata always guaranteed to be the last chunk?

File: ddtrace/contrib/internal/langchain/patch.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1082]: Let's avoid the triple nesting here

File: ddtrace/contrib/internal/langchain/patch.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1105]: Does LLM.stream() also accept multiple types of inputs that we need to account for?

File: ddtrace/contrib/internal/langchain/patch.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1116]: We can probably reduce indentation here too

File: ddtrace/contrib/internal/langchain/patch.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1119]: Same here

File: ddtrace/contrib/internal/langchain/patch.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1368]: I assume these are due to import errors on langchain's end (not being able to directly access them without importing first?)

File: ddtrace/contrib/internal/langchain/patch.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1392]:
```suggestion
```

File: ddtrace/contrib/internal/langchain/utils.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/utils.py:11]: Looking at langchain's [source code](https://github.com/langchain-ai/langchain/blob/63c3cc1f1f2b78eb588e56fb62a7dacb43a2e27f/libs/core/langchain_core/language_models/chat_models.py#L408-L422), it looks like they actually add the chunks for us in the stream() operation. Would it be possible for us to access/reuse that instead of performing the appending/processing ourselves?

File: ddtrace/contrib/internal/langchain/utils.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/utils.py:40]: This is part of a larger refactor I'm pushing for, but can we get rid of the `error` arg in these handlers? This is already captured as part of `span.error` so I don't see why we need this to be passed explicitly

File: tests/contrib/langchain/conftest.py
Comment (Yun-Kim) [inline tests/contrib/langchain/conftest.py:169]: Let's remove TODOs in code and move them to internal tickets instead

File: tests/contrib/langchain/test_langchain_community.py
Comment (Yun-Kim) [inline tests/contrib/langchain/test_langchain_community.py:1287]: What does the chat/completions client_path do?

**Review by Yun-Kim — COMMENTED** [review]
Great work! I added a few comments about styling (avoiding nesting when possible for both readability and ironically line count), and some clarification questions but this is great. Awesome job on setting up a test fixture instead of being bogged down by vcrpy!

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py:986]: good question, let me explore that more. since we're not concerned with older versions of langchain/non-LCEL, i'll likely ignore those and verify that the name `input` is consistent and that it shows as the first in `args`

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py:1009]: it's a bit tricky with chains, because there could be an output parser that _does_ turn the chunks into strings, or they could be straight up `BaseMessage` types (ie `AIMessage`). in our current LCEL `chain` patching, we just [stringify](https://github.com/DataDog/dd-trace-py/blob/baac738726a6b615cb3f79693b0b1f117b688f9d/ddtrace/contrib/internal/langchain/patch.py#L844) to avoid going case-by-case.

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py:1064]: oop good catch thanks!

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py:1079]: I believe so (if it's included), but I'll try and look to see if there's evidence it could show up in a different chunk

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py:1105]: ah yes good catch, it actually does look like it accepts a union of different values (although, different from the set accepted by chat models). will update to more accurately reflect that!

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py:1368]: tbh i just added them by default. let me try and remove them and see what happens

File: ddtrace/contrib/internal/langchain/utils.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/utils.py:11]: good find! that one looks specific to chat models, i'll check if their runnable sequence (chain) implements something similar.

File: ddtrace/contrib/internal/langchain/utils.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/utils.py:40]: yep can do

File: tests/contrib/langchain/test_langchain_community.py
Comment (sabrenner) [reply tests/contrib/langchain/test_langchain_community.py:1287]: accesses the `openai.chat.completion` path specifically to set as the client to use for the `langchain_openai.ChatOpenAI` or `langchain_openai.OpenAI`. had to look for [this line](https://github.com/langchain-ai/langchain/blob/63c3cc1f1f2b78eb588e56fb62a7dacb43a2e27f/libs/partners/openai/langchain_openai/chat_models/base.py#L516-L517) to verify this is what kind of client they were setting. i can add a comment clarifying this

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py:979]: I will leave a note in the description of your "refactor langchain ticket" for both of the TODOs to make sure they get addressed 😄

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py]: yep that makes sense to me! tried it out and it worked. nice find! (also added checks for `langchain_core`)

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py:1364]: looks like it works without them 🫡

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py]: this might be openai-specific. since we already have a check in the non-streamed chat model patching to check `_is_openai_chat_instance`, i'll add that here as well, and in that case I believe it is guaranteed to be on the last chunk

File: ddtrace/contrib/internal/langchain/utils.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/utils.py:11]: it looks like it would only be accessible through the `run_manager` instance, which relies on the callback configs. i think we've thought about patching callback configs to run our own callbacks (which would good for handling these kinds of events for span management), but i think it's a bit out of scope for this PR. otherwise, it doesn't look like that accumulated `generations` goes anywhere we'd be able to use it 😢

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py:985]: yep, it looks like this named argument has been in place since the end of the `0.0.x` line, so we should be good here.

File: ddtrace/contrib/internal/langchain/utils.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/utils.py:86]: Can we add a follow up item to submit llmobs spans from streamed responses?

File: ddtrace/contrib/internal/langchain/patch.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1050]: Just to be safe, is `chunk` guaranteed to have the `content` attribute?

File: ddtrace/contrib/internal/langchain/patch.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1054]: should also return if `usage` is not a dict type (would break in 1056 otherwise)

File: ddtrace/contrib/internal/langchain/patch.py
Comment (Yun-Kim) [reply ddtrace/contrib/internal/langchain/patch.py:1082]: meh, as long as we're not making up roles when they don't exist in `llm.generate()` cases, I'd rather push for a unified way of tagging llms vs chat models, i.e. treat them all as messages. I think this is fine

File: ddtrace/contrib/internal/langchain/patch.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1094]: Just for clarity: llm streams chunks are in the form of raw text?

File: ddtrace/contrib/internal/langchain/patch.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1206]: Removing this didn't break patching?

File: tests/contrib/langchain/test_langchain_community.py
Comment (Yun-Kim) [inline tests/contrib/langchain/test_langchain_community.py:1426]: So to confirm, we do not trace astream_events() (there is also no `stream_events()`, just `astream_events()`?), but it does in fact interfere with `stream()` so possibly can be affected by our patching of `stream()`?

Could we add a note in the PR description and a JIRA item to follow up on this?

File: ddtrace/contrib/internal/langchain/utils.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/utils.py:5]: We don't need to base this off `wrapt.ObjectProxy`?

File: ddtrace/contrib/internal/langchain/utils.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/utils.py:118]: Let's only attach role if role is not None, in case of llm.stream() being called rather than chat_model.

File: releasenotes/notes/langchain-lcel-stream-calls-bff85c974a72cceb.yaml
Comment (Yun-Kim) [inline releasenotes/notes/langchain-lcel-stream-calls-bff85c974a72cceb.yaml:5]:
```suggestion
    Note that due to an upstream issue with the ``langchain`` library itself, streamed responses will not be tagged correctly based on the choice index when the underlying model is configured to return ``n>1`` choices.
```
Can we also link the associated github issue in this release note entry?

**Review by Yun-Kim — COMMENTED** [review]
Looks good, mostly clarification questions! For easier tracking, can we add a limitations section to this PR description (i.e. n>1 response tagging, astream_events())? Thanks!

File: ddtrace/contrib/internal/langchain/utils.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/utils.py:86]: yep i have a ticket on our board!

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py:1050]: yeah, `model.stream` for any generic `model` from a langchain partner library uses the base definition of `stream` in `langchain_core`, which returns an iterator of `BaseMessage` chunks, which have `content` on them. But I agree it's best to be safe, I'll use `getattr` with a default of the chunk itself and stringify

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py:1054]: ah yes good catch!

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py:1094]: similar to the chat model case, yeah their `llm.stream` definition specifies it's an [iterator of strings](https://github.com/langchain-ai/langchain/blob/f9ef688b3a0e3f576a76435e242090a76666974b/libs/core/langchain_core/language_models/llms.py#L515-L522).

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py:1206]: ah i didn't mean to remove this line, thanks. but, it looks like all tests still passed. i'll put it back since it's out of scope for this PR, but maybe we can try removing it in a different one if we don't need it here

File: tests/contrib/langchain/test_langchain_community.py
Comment (sabrenner) [reply tests/contrib/langchain/test_langchain_community.py:1426]: yep that's my understanding of how it works, and yes no `stream_events`. I'll make a note + follow up ticket

File: ddtrace/contrib/internal/langchain/utils.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/utils.py:5]: i had tried that initially, as i took inspiration from the antropic integration. but, when doing that, i end up getting
```
AttributeError: 'generator' object has no attribute '_dd_integration'
```
so i think this is why i went with this approach

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py]: added a safety check in https://github.com/DataDog/dd-trace-py/pull/10672/commits/47a3f15fdc43b840d42c6749b0a9ed2681c68a19

File: ddtrace/contrib/internal/langchain/patch.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1051]: Do we not want to tag role from the output messages as well?

File: ddtrace/contrib/internal/langchain/patch.py
Comment (Yun-Kim) [inline ddtrace/contrib/internal/langchain/patch.py:1014]: I think we can rename the tag name to `langchain.response.outputs` since this is the output of the chain, and not necessarily a llm/chat model, right?

**Review by Yun-Kim — APPROVED** [review]
Two last questions, I'll approve to unblock you but please address them (I'll defer to your judgement). Thanks!

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py:1051]: ah yes good call

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py:1014]: yeah that makes sense to me!

File: ddtrace/contrib/internal/langchain/patch.py
Comment (sabrenner) [reply ddtrace/contrib/internal/langchain/patch.py:1052]: tagged it as `message_type` to be consistent with our normal `generate` patching.

**Pull request merged**
